docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude

```diff
@@ -0,0 +1,113 @@
+            var dataPoints = GenerateRandomDataPoints(1000, <#=DataSeed#>);
+
+            // Convert the list of data points to an IDataView object, which is consumable by ML.NET API.
+            var trainingData = mlContext.Data.LoadFromEnumerable(dataPoints);
+
+<# if (TrainerOptions == null) { #>
+            // Define the trainer.
+            var pipeline = mlContext.Clustering.Trainers.<#=Trainer#>(<#=InlineTrainerOptions#>);
+<# } else { #>
+            // Define trainer options.
+            var options = new <#=TrainerOptions#>;
+
+            // Define the trainer.
+            var pipeline = mlContext.Clustering.Trainers.<#=Trainer#>(options);
+<# } #>
+
+            // Train the model.
+            var model = pipeline.Fit(trainingData);
+
+            // Create testing data. Use different random seed to make it different from training data.
+            var testData = mlContext.Data.LoadFromEnumerable(GenerateRandomDataPoints(500, seed: 123));
+
+            // Run the model on test data set.
+            var transformedTestData = model.Transform(testData);
+
+            // Convert IDataView object to a list.
+            var predictions = mlContext.Data.CreateEnumerable<Prediction>(transformedTestData, reuseRowObject: false).ToList();
+
+            // Look at 5 predictions
+            foreach (var p in predictions.Take(2))
+                Console.WriteLine($"Label: {p.Label}, Prediction: {p.PredictedLabel}");
+            foreach (var p in predictions.TakeLast(3))
+                Console.WriteLine($"Label: {p.Label}, Prediction: {p.PredictedLabel}");
+
+            <#=ExpectedOutputPerInstance#>
+
+            // Evaluate the overall metrics
+            var metrics = mlContext.Clustering.Evaluate(transformedTestData, "Label", "Score", "Features");
+            Console.WriteLine($"Normalized Mutual Information: {metrics.NormalizedMutualInformation:F2}");
+            Console.WriteLine($"Average Distance: {metrics.AverageDistance:F2}");
+            Console.WriteLine($"Davies Bouldin Index: {metrics.DaviesBouldinIndex:F2}");
+            
+            <#=ExpectedOutput#>
+
+            // Get cluster centroids and the number of clusters k from KMeansModelParameters.
+            VBuffer<float>[] centroids = default;
+
+            var modelParams = model.Model;
+            modelParams.GetClusterCentroids(ref centroids, out int k);
+            Console.WriteLine($"The first 3 coordinates of the first centroid are: ({string.Join(", ", centroids[0].GetValues().ToArray().Take(3))})");
+            Console.WriteLine($"The first 3 coordinates of the second centroid are: ({string.Join(", ", centroids[1].GetValues().ToArray().Take(3))})");
+
+            <#=ExpectedCentroidsOutput#>
+        }
+
+        private static IEnumerable<DataPoint> GenerateRandomDataPoints(int count, int seed = 0)
+        {
+            var random = new Random(seed);
+            float randomFloat() => (float)random.NextDouble();
```

Comment:
>(float)random.NextDouble() [](start = 35, length = 26)

- 0.5f

Reply:
I don't think that centering at 0 will give any significant advantage. 

---
In reply to: [275080652](https://github.com/dotnet/machinelearning/pull/3317#discussion_r275080652) [](ancestors = 275080652)